core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -347,19 +347,119 @@ public static String sha512(ByteString string)  {
+   *  Returns NULL if there is no match. Returns an exception if regex is invalid.
+   *  Uses position=1 and occurrence=1 as default values when not specified. */
+  public static @Nullable String regexpExtract(String value, String regex) {
+    return regexpExtract(value, regex, 1, 1);
+  }
+
+  /** SQL {@code REGEXP_EXTRACT(value, regexp, position)} function.
+   *  Returns NULL if there is no match, or if position is beyond range.
+   *  Returns an exception if regex or position is invalid.
+   *  Uses occurrence=1 as default value when not specified. */
+  public static @Nullable String regexpExtract(String value, String regex, int position) {
+    return regexpExtract(value, regex, position, 1);
+  }
+
+  /** SQL {@code REGEXP_EXTRACT(value, regexp, position, occurrence)} function.
+   *  Returns NULL if there is no match, or if position or occurrence are beyond range.
+   *  Returns an exception if regex, position or occurrence are invalid. */
+  public static @Nullable String regexpExtract(String value, String regex, int position,
+      int occurrence) {
+    // Uses java.util.regex as a standard for regex processing
+    // in Calcite instead of RE2 used by BigQuery/GoogleSQL
+    Pattern regexp;
+    String methodName = "REGEXP_EXTRACT";
+    try {
+      regexp = Pattern.compile(regex);
+    } catch (PatternSyntaxException ex) {
+      throw RESOURCE.invalidRegexInputForRegexpFunctions(ex.getMessage()
+          .replace(System.lineSeparator(), " "), methodName).ex();
+    }
+
+    Matcher matcher = regexp.matcher(value);
+
+    if (position <= 0) {
+      throw RESOURCE.invalidIntegerInputForRegexpFunctions(Integer.toString(position),
+          "position", methodName).ex();
+    }
+    if (position > value.length()) {
+      return null;
+    }
+    matcher.region(position - 1, value.length());
+
+    if (occurrence <= 0) {
+      throw RESOURCE.invalidIntegerInputForRegexpFunctions(Integer.toString(occurrence),
+          "occurrence", methodName).ex();
+    }
+
+    if (matcher.groupCount() > 1) {
+      throw RESOURCE.multipleCapturingGroupsForRegexpExtract(
+          Integer.toString(matcher.groupCount()), methodName).ex();
+    }
+
+    String match = null;
+    while (occurrence > 0) {
+      if (matcher.find()) {
+        if (matcher.groupCount() == 1) {
+          match = matcher.group(1);
+        } else {
+          match = matcher.group(0);
+        }
+      } else {
+        return null;
+      }
+      occurrence--;
+    }
+
+    return match;
+  }
+
+  /** SQL {@code REGEXP_EXTRACT_ALL(value, regexp)} function.
+   *  Returns an empty array if there is no match, returns an exception if regex is invalid.*/
+  public static List<String> regexpExtractAll(String value, String regex) {
+    // Uses java.util.regex as a standard for regex processing
+    // in Calcite instead of RE2 used by BigQuery/GoogleSQL
+    Pattern regexp;
+    String methodName = "REGEXP_EXTRACT_ALL";
```

Comment:
REGEXP_EXTRACT and REGEXP_EXTRACT_ALL are pretty similar. What do you think of anytime REGEXP_EXTRACT gets called, you can call REGEXP_EXTRACT_ALL method and then just retrieve the requested element (e.g. if the occurrence is 2 then retrieve the second element). Would this simplify things?

Reply:
That makes sense @tanclary, I considered that approach but REGEXP_EXTRACT_ALL does not have the position or occurrence params like REGEXP_EXTRACT, so merging the functionality as you suggested would mean we find all occurrences of the pattern for every call to EXTRACT and then picking the required index (also these are optional params so many a times we would be only looking for the default first occurrence). This can be avoided by passing the occurrence to EXTRACT_ALL and stopping the loop on reaching that count, but we'd have to include some logic to skip this when we need the whole list of matches (extract_all). I felt this might be added complexity with some hit on the performance, hence decoupled it even though they have similar functionality. What do you think is it better to consolidate them?

Reply:
That makes sense. I just noticed some of the logic is repeated between the functions so maybe at the least we can extract that into a helper or something. We should try to avoid repeated checks/logic across these methods if we can.

Reply:
Hi @tanclary, moved the redundant error checks into a couple of helper functions to reuse majority of the code, I've kept the matching logic separate due to the above reasons would be good if you can have a look and suggest any improvements if needed.

Reply:
That sounds good, once we get REGEXP_EXTRACT merged and can rebase this I'll come back for a second review 👍

Reply:
@tanclary I've rebased the changes from previous PRs, also consolidated few of the redundant error checks into helper methods would be good if you can give it a review!